### DISCUSSION TITLE
Kubernetes中Sidecar生命周期管理 - Qinng

### DISCUSSION THREAD

**Discussion by giscus** [qingwave/qingwave.github.io#80] [Announcements]
# Kubernetes中Sidecar生命周期管理 - Qinng

背景在多个容器的Pod中，通常业务容器需要依赖sidecar。启动时sidecar需要先启动，退出时sidecar需要在业务容器退出后再退出。k8s目前对于sidecar的生命周期比较有争议，见iss

https://qingwave.github.io/k8s-sideccar-lifecycle/

**Comment by JianXinyu** [comment]
这 main container 并不会自己结束呀。

**Reply by qingwave** [reply]
很多场景下main容器是会主动或被动的退出，比如升级, 异常退出，节点滚动更新

**Reply by JianXinyu** [reply]
真实场景下是这样的。只是这个例子中你是咋结束main container的？除此一点疑问之外，非常棒的讲解。

**Reply by qingwave** [reply]
我们不需要自己结束main，只需要在main退出时sidecar能感知到，再退出sidecar来保证服务质量，例子中是在main启动时包裹了trap监听信号，当main退出时会创建一个文件，sidecar通过prestop直到监测到文件存在才会主动退出。

当然，也还有其他方式，在prestop监测main端口也是可行的

**Reply by JianXinyu** [reply]
感谢回复。只是我在测试这个例子时，main 会困在 until 循环里一直打印日期，不会结束了。

**Reply by qingwave** [reply]
哦哦哦，会错意了，这边是模拟长运行时的业务，直接kubectl delete模拟删除时即可

**Reply by JianXinyu** [reply]
哈哈我想也是，感谢！
